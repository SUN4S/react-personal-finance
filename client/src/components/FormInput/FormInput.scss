label {
  width: 100%;
  color: var(--text-primary);
  margin-top: 24px;

  input,
  textarea,
  select {
    height: 40px;
    width: 100%;
    font-size: 16px;
    padding-left: 6px;
    margin-top: 3px;
    margin-bottom: 1px;
    border: 2px solid transparent;
    border-radius: 3px;
    color: var(--text-primary);
    background-color: var(--sidebar-bg);
    outline: none;
    transition: all 0.3s ease;
    background-color: var(--page-bg);
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;

    &:focus {
      border: 2px solid var(--primary);
      opacity: 0.9;
    }
  }
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type="number"] {
  -moz-appearance: textfield;
}
