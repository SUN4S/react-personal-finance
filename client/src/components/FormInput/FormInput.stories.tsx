import { Meta, Story } from "@storybook/react/types-6-0";

import { FormInput } from "./FormInput";
import { FormInputProps } from "../../models/inputs";
import { useForm } from "react-hook-form";

interface TemplateProps extends FormInputProps {
  theme: string;
}

export default {
  title: "Components/Form Input",
  component: FormInput,
} as Meta;

const Template: Story<TemplateProps> = (args) => {
  const { register } = useForm();
  return (
    <html data-theme={args.theme}>
      <body
        style={{
          width: "400px",
          backgroundColor: "var(--sidebar-bg)",
          padding: "8px",
        }}
      >
        <FormInput
          labelFor={args.labelFor}
          label={args.label}
          name={args.name}
          inputTestId={args.inputTestId}
          required={args.required}
          type={args.type}
          placeholder={args.placeholder}
          register={register}
        />
      </body>
    </html>
  );
};

export const Default_Light = Template.bind({});
Default_Light.args = {
  theme: "light",
  labelFor: "name",
  label: "Enter your name: ",
  name: "name",
  inputTestId: "input",
  required: true,
  type: "text",
  step: 0.1,
  placeholder: "Your name here",
};

export const Default_Dark = Template.bind({});
Default_Dark.args = {
  theme: "dark",
  labelFor: "name",
  label: "Enter your name: ",
  name: "name",
  inputTestId: "input",
  required: true,
  type: "text",
  step: 0.1,
  placeholder: "Your name here",
};
