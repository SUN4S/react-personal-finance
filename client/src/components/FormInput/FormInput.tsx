import "./FormInput.scss";

import { FieldErrors, FieldValues } from "react-hook-form";

import { Children } from "react";

type InputType = "text" | "password" | "email" | "number";

interface InputProps {
  labelFor: string;
  label: string;
  name?: string;
  required?: boolean;
  type?: InputType;
  step?: number;
  placeholder?: string;
  register?: any;
  children?: JSX.Element;
}

export const FormInput: React.FC<InputProps> = ({
  labelFor,
  label,
  name,
  required,
  type,
  step,
  placeholder,
  register,
  children,
}) => {
  return (
    <label htmlFor={labelFor}>
      {label}
      {children ? (
        children
      ) : (
        <input
          type={type}
          step={step}
          {...register(name, { required: required })}
          placeholder={placeholder}
        />
      )}
    </label>
  );
};
