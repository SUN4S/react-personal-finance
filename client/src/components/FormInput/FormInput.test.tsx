import { cleanup, fireEvent, render, screen } from "@testing-library/react";

import { FormInput } from "./FormInput";
import { FormProvider } from "react-hook-form";
import renderer from "react-test-renderer";

describe("Form Input", () => {
  afterEach(cleanup);
  const mockRegister = jest.fn();
  it("Renders", () => {
    const { getByTestId } = render(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    expect(getByTestId("formLabel")).toBeTruthy();
  });

  it("Render Label", () => {
    const { getByTestId } = render(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    expect(getByTestId("formLabel")).toHaveTextContent("name");
  });

  it("Render Input", () => {
    const { getByTestId } = render(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    expect(getByTestId("formInput")).toBeTruthy();
  });

  it("Renders Placeholder", () => {
    const { getByTestId } = render(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    const inputPlaceholder = screen.getByPlaceholderText("Enter Name");
    expect(inputPlaceholder).toBeTruthy();
  });

  it("Is Typeable", () => {
    const { getByTestId } = render(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    const input = screen.getByTestId("formInput");
    fireEvent.change(input, { target: { value: "John" } });
    expect(input).toHaveValue("John");
  });

  it("Renders Children", () => {
    const { getByTestId } = render(
      <FormInput labelFor="image">
        <h1>Hello World</h1>
      </FormInput>
    );
    expect(getByTestId("formLabel")).toContainHTML("<h1>Hello World</h1>");
  });

  it("Matches Snapshot", () => {
    const tree = renderer.create(
      <FormInput
        labelFor="name"
        label="name"
        required
        type="text"
        placeholder="Enter Name"
        register={mockRegister}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it("Matches Snapshot (Children)", () => {
    const tree = renderer.create(
      <FormInput labelFor="image">
        <h1>Hello World</h1>
      </FormInput>
    );
    expect(tree).toMatchSnapshot();
  });
});
