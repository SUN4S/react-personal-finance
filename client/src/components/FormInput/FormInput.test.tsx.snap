// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Input Matches Snapshot (Children) 1`] = `
<label
  data-testid="formLabel"
  htmlFor="image"
>
  <h1>
    Hello World
  </h1>
</label>
`;

exports[`Form Input Matches Snapshot 1`] = `
<label
  data-testid="formLabel"
  htmlFor="name"
>
  name
  <input
    data-testid="formInput"
    placeholder="Enter Name"
    type="text"
  />
</label>
`;
